h/services/bulk_annotation.py

```diff
@@ -0,0 +1,190 @@
+            "lte": "2012-11-30",
+            "eq": "2012-11-30",
+            "ne": "2012-11-30",
+        }
+
+    :raises BadDateFilter: For unrecognised operators or no spec
+    """
+    if not spec:
+        raise BadDateFilter(f"No spec given to filter '{column}' on")
+
+    clauses = []
+
+    for op_key, value in spec.items():
+        if op_key == "gt":
+            clauses.append(column > value)
+        elif op_key == "gte":
+            clauses.append(column >= value)
+        elif op_key == "lt":
+            clauses.append(column < value)
+        elif op_key == "lte":
+            clauses.append(column <= value)
+        elif op_key == "eq":
+            clauses.append(column == value)
+        elif op_key == "ne":
+            clauses.append(column != value)
+        else:
+            raise BadDateFilter(f"Unknown date filter operator: {op_key}")
+
+    return sa.and_(*clauses)
+
+
+class BulkAnnotationService:
+    """A service for retrieving annotations for many users at once."""
+
+    # Aliases to distinguish between different types of user
+    _AUTHOR = sa.orm.aliased(User, name="author")
+    _AUDIENCE = sa.orm.aliased(User, name="audience")
+
+    # Acceptable values to pass in as `fields`
+    _FIELDS = {
+        "author.username": _AUTHOR.username,
+        "group.authority_provided_id": Group.authority_provided_id,
+    }
+
+    def __init__(self, db_session: Session):
+        """Initialise the service."""
+
+        self._db = db_session
+
+    # pylint: disable=too-many-arguments
+    def annotation_search(
+        self,
+        authority: str,
+        audience: dict,
+        updated: dict,
+        fields: list = None,
+        limit=100000,
+    ) -> Union[List[Annotation], Row]:
+        """
+        Get a list of annotations / rows viewable by a particular set of users.
+
+        Using a fields argument will switch the return type from annotation
+        objects to row objects. This is more efficient, but less convenient if
+        you want many fields. Check the `_FIELDS` attribute for acceptable
+        fields and names.
+
+        :param authority: The authority to search by
+        :param audience: A specification of how to find the users. e.g.
+            {"username": [...]}
+        :param updated: A specification of how to filter the updated date. e.g.
+            {"gt": "2019-01-20", "lte": "2019-01-21"}
+        :param fields: A list of string descriptions of fields to return
+            instead of full annotation objects.
+        :param limit: A limit of results to generate
+        :raises BadDateFilter: For poorly specified date conditions
+        :raises BadFieldSpec: For poorly specified fields
+        """
+
+        results = self._db.execute(
+            self._search_query(
+                authority, audience=audience, updated=updated, fields=fields
+            ).limit(limit)
+        )
+
+        if fields is None:
+            results = results.scalars()
```

Comment:
We allow the caller to call without fields. You don't need this for this exact use case, but it's very likely you'd want it in other use cases.

It also happens to be _very_ handy for testing. We can test the annotations much more naturally, and then separately check the field formatting.